crates/rpc/rpc-eth-types/src/gas_oracle.rs

```diff
@@ -340,11 +341,148 @@ where
+    /// gas would exceed the block's gas limit, or the total block payload size plus the maximum
+    /// single transaction would exceed the block's payload size limit.
+    pub async fn scroll_suggest_tip_cap(
+        &self,
+        min_suggested_priority_fee: U256,
+        payload_size_limit: u64,
+    ) -> EthResult<U256> {
+        let (result, _) = self
+            .calculate_suggest_tip_cap(
+                BlockNumberOrTag::Latest,
+                min_suggested_priority_fee,
+                payload_size_limit,
+            )
+            .await;
+        result
+    }
+
+    /// Calculates a gas price suggestion and returns whether the block is at capacity.
+    ///
+    /// This method implements the core logic for suggesting gas prices based on block capacity.
+    /// It returns a tuple containing the suggested gas price and a boolean indicating
+    /// whether the latest block is at capacity (gas limit or payload size limit).
+    pub async fn calculate_suggest_tip_cap(
+        &self,
+        block_number_or_tag: BlockNumberOrTag,
+        min_suggested_priority_fee: U256,
+        payload_size_limit: u64,
+    ) -> (EthResult<U256>, bool) {
+        let header = match self.provider.sealed_header_by_number_or_tag(block_number_or_tag) {
+            Ok(Some(header)) => header,
+            Ok(None) => return (Err(EthApiError::HeaderNotFound(BlockId::latest())), false),
+            Err(e) => return (Err(e.into()), false),
+        };
+
+        let mut inner = self.inner.lock().await;
+
+        // if we have stored a last price, then we check whether or not it was for the same head
+        if inner.last_price.block_hash == header.hash() {
+            return (Ok(inner.last_price.price), inner.last_price.is_at_capacity);
+        }
+
+        let mut suggestion = min_suggested_priority_fee;
+        let mut is_at_capacity = false;
+
+        // find the maximum gas used by any of the transactions in the block and
+        // the maximum and total payload size used by the transactions in the block to use as
+        // the capacity margin for the block, if no receipts or block are found return the
+        // suggested_min_priority_fee
+        let (block, receipts) = match self.cache.get_block_and_receipts(header.hash()).await {
+            Ok(Some((block, receipts))) => (block, receipts),
+            Ok(None) => return (Ok(suggestion), false),
+            Err(e) => return (Err(e.into()), false),
+        };
+
+        let max_tx_gas_used = match receipts.len() {
+            // If no receipts, return the default suggestion
+            0 => return (Ok(suggestion), is_at_capacity),
+            // If only one receipt, use its cumulative gas (which equals the gas used by that
+            // transaction)
+            1 => receipts[0].cumulative_gas_used(),
+            // If multiple receipts, calculate individual transaction gas usage
+            _ => receipts
```

Comment:
nit (feel free to skip if you don't like the syntax):
```suggestion
        let max_tx_gas_used = match receipts.as_slice() {
            // If no receipts, return the default suggestion
            [] => return (Ok(suggestion), is_at_capacity),
            // If only one receipt, use its cumulative gas (which equals the gas used by that
            // transaction)
            [ref receipt] => receipt.cumulative_gas_used(),
            // If multiple receipts, calculate individual transaction gas usage
            _ => receipts
```

Reply:
updated this logic